twilight-http-ratelimiting/src/actor.rs

```diff
@@ -0,0 +1,380 @@
+//! Rate limiting state manager.
+
+use crate::{Bucket, Path, Predicate, RateLimitHeaders, GLOBAL_LIMIT_PERIOD};
+use hashbrown::{hash_table::Entry as TableEntry, HashTable};
+use std::{
+    collections::{hash_map::Entry as MapEntry, HashMap, VecDeque},
+    future::poll_fn,
+    hash::{BuildHasher, Hash, Hasher as _, RandomState},
+    mem,
+    pin::pin,
+};
+use tokio::{
+    sync::{
+        mpsc,
+        oneshot::{self, error::RecvError},
+    },
+    task::JoinSet,
+    time::{sleep, Duration, Instant},
+};
+use tokio_util::time::delay_queue::{DelayQueue, Key};
+
+/// Rate limiter hasher.
+#[derive(Debug)]
+struct Hasher(RandomState);
+
+impl Hasher {
+    /// Hashes a path.
+    fn hash(&self, path: &Path) -> u64 {
+        let mut hasher = self.0.build_hasher();
+        path.hash_components(&mut hasher);
+        hasher.finish()
+    }
+    /// Hashes a bucket and a path.
+    fn hash_bucket(&self, bucket: &[u8], path: &Path) -> u64 {
+        let mut hasher = self.0.build_hasher();
+        path.hash_components(&mut hasher);
+        bucket.hash(&mut hasher);
+        hasher.finish()
+    }
+}
+
+/// Pending permit request state.
+#[derive(Debug)]
+pub struct Message {
+    /// Completion handle.
+    pub notifier: oneshot::Sender<oneshot::Sender<Option<RateLimitHeaders>>>,
+    /// Path the permit is for, mapping to a [`Queue`].
+    pub path: Path,
+}
+
+/// Grouped pending permits holder.
+///
+/// Grouping may be done by path or bucket, based on previous permits' response
+/// headers.
+///
+/// Queue may not be rate limited, in which case the values of [`limit`][Self::limit],
+/// [`reset`][Self::reset], and [`remaining`][Self::remaining] are unused.
+#[derive(Debug)]
+struct Queue {
+    /// Whether the queue is handling outstanding permits.
+    ///
+    /// Note that this is `true` when globally exhausted and `false` when
+    /// the queue is exhausted.
+    idle: bool,
+    /// List of pending permit requests.
+    inner: VecDeque<Message>,
+    /// Total number of permits until the queue becomes exhausted.
+    limit: u16,
+    /// Key mapping to an [`Instant`] when the queue resets, if rate limited.
+    reset: Option<Key>,
+    /// Number of remaining permits until the queue becomes exhausted.
+    remaining: u16,
+}
+
+impl Default for Queue {
+    fn default() -> Self {
+        Self {
+            idle: true,
+            inner: VecDeque::new(),
+            limit: 0,
+            reset: None,
+            remaining: 0,
+        }
+    }
+}
+
+/// Interval at which to prune stale queues.
+const GC_INTERVAL: Duration = Duration::from_secs(60 * 60 * 6);
+
+/// Rate limiter actor runner.
+#[allow(clippy::too_many_lines)]
+pub async fn runner(
+    global_limit: u16,
+    mut rx: mpsc::UnboundedReceiver<(Message, Option<Predicate>)>,
```

Comment:
I'd prefer a new struct here to avoid `.0` and `.1` and for more clarity

Reply:
I've replaced most `.0` and `.1` with tuple destructuring. PTAL